rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -126,17 +132,208 @@ void Player::play_messages_from_queue()
   } while (!is_storage_completely_loaded() && rclcpp::ok());
 }
 
+//Dynamic Alignment as Fastrtps
+unsigned long alignment(unsigned long data_size, unsigned long last_data_size, unsigned long current_position)
+{
+  return data_size > last_data_size ? (data_size - current_position % data_size) & (data_size-1):0;
+}
+
+//deal with string
+void Player::deal_with_string(const uint8_t *dds_buffer, bool is_wstring)
+{
+  uint32_t length;
+  size_t string_header = sizeof(uint32_t);//string header 4 Bytes
+
+  unsigned long one_offset = alignment(string_header, last_data_size, current_position);
+  current_position = current_position + one_offset;
+  memcpy(&length, (dds_buffer + current_position + 4), string_header);
+
+  if(!is_wstring)
+    {
+      last_data_size = sizeof(char);
+      current_position += (string_header + length);
+    }
+  else {
+      last_data_size = sizeof(uint32_t);
+      current_position += (string_header + length * sizeof(uint32_t));
+  }
+}
+
+//find out the real position of header in fastrtps serialized data
+void Player::calculate_position_with_align(const uint8_t * dds_buffer_ptr, const rosidl_typesupport_introspection_cpp::MessageMember *message_member, unsigned long stop_index)
```

Comment:
I think this change should actually land in `rosidl_typesupport_introspection_cpp` rather than in rosbag2.
We've worked on a similar change for the serialized bag message, which resulted in here: https://github.com/ros2/rosidl/pull/416

So I am very hesitant to bring back this logic

Reply:
I can confirm that this is probably not the right place for a lot of this logic. the point is that we don't really know if this PR is even necessary. There is currently no way (as far as we know) of achieving what #299 needs without compromising the relatively clean rosbag functionality. So this is just the first shot we have for this problem so any suggestion is helpful :)

Reply:
@Karsten1987  the ros2/rosidl#416 you mentioned does not solve the problem. We do not know the type at hand (we cannot static cast something unknown into something unknown) ... 
``` std_msgs__msg__String msg = *static_cast<std_msgs__msg__String *>(message_memory); ```
in your solution does not help when dealing with fully unknown type deserialization. This could only work if we ```switch(type_string) case 1, case 2, case 3``` all possible message types - which is not possible due to the existence of custom messages (which are not known by default).

This works well for subscribers (where I give the expected Template type of the deserialization) ... but it should not work in places where i deserialize something random and look for fields.

Am I missing something here? Thanks :)